src/controller.cpp

```diff
@@ -241,29 +303,37 @@ void Controller::connected() {
 void Controller::connectionConfirmed() {
   logger.log() << "Connection confirmed";
 
-  if (m_state != StateConfirming) {
+  if (m_state != StateConfirming && m_state != StateOn) {
     logger.log() << "Invalid confirmation received";
     return;
   }
 
   m_connectionRetry = 0;
   emit connectionRetryChanged();
 
-  setState(StateOn);
-  emit timeChanged();
+  if (m_state != StateOn) {
+    setState(StateOn);
+    emit timeChanged();
 
-  if (m_nextStep != None) {
-    deactivate();
-    return;
-  }
+    if (m_nextStep != None) {
+      deactivate();
+      return;
+    }
 
-  m_timer.start(TIMER_MSEC);
+    m_timer.start(TIMER_MSEC);
+  } else {
+    emit silentSwitchDone();
+  }
 }
 
 void Controller::connectionFailed() {
   logger.log() << "Connection failed!";
 
-  if (m_state != StateConfirming) {
+  if (m_state == StateOn) {
```

Comment:
Move this block after line 336.

Reply:
Plus, if we are unable to connect, we are in a bad scenario. Maybe we should call `deactivate()`. Or have a connectionRetry logic. What do you think?

Reply:
I did not call `deactivate()` to use the connectionRetry logic from the function. 
If calls `m_impl->deactivate(ControllerImpl::ReasonConfirming)` which triggers the retry. 

What we could do that would improve the logic is that after unsuccessful retries we silently switch servers again. This however would be a little bit more complex but would ensure that we find the best server.